py/picca/delta_extraction/data_catalogues/desi_data.py

```diff
@@ -311,78 +334,162 @@ def format_data(self,
-                    "dec": row['DEC'],
-                    "z": row['Z'],
-                }
-                args["log_lambda"] = np.log10(spec['WAVELENGTH'])
-
-                if self.analysis_type == "BAO 3D":
-                    forest = DesiForest(**args)
-                elif self.analysis_type == "PK 1D":
-                    if self.use_non_coadded_spectra:
-                        exposures_diff = exp_diff_desi(spec, w_t)
-                        if exposures_diff is None:
-                            continue
-                    else:
-                        exposures_diff = np.zeros(spec['WAVELENGTH'].shape)
-                    if reso_from_truth:
-                        reso_sum = spec['RESO'][:, :]
-                    else:
-                        if len(spec['RESO'][w_t].shape) < 3:
-                            reso_sum = spec['RESO'][w_t].copy()
-                        else:
-                            reso_sum = spec['RESO'][w_t].sum(axis=0)
-                    reso_in_pix, reso_in_km_per_s = spectral_resolution_desi(
-                        reso_sum, spec['WAVELENGTH'])
-                    args["exposures_diff"] = exposures_diff
-                    args["reso"] = reso_in_km_per_s
-                    args["resolution_matrix"] = reso_sum
-                    args["reso_pix"] = reso_in_pix
-
-                    forest = DesiPk1dForest(**args)
-                # this should never be entered added here in case at some point
-                # we add another analysis type
-                else:  # pragma: no cover
-                    raise DataError("Unkown analysis type. Expected 'BAO 3D'"
-                                    f"or 'PK 1D'. Found '{self.analysis_type}'")
-
-                # rebin arrays
-                # this needs to happen after all arrays are initialized by
-                # Forest constructor
-                forest.rebin()
-
-                # keep the forest
-                if targetid in forests_by_targetid:
-                    existing_forest = forests_by_targetid[targetid]
-                    existing_forest.coadd(forest)
-                    forests_by_targetid[targetid] = existing_forest
+                # If keep_single_exposures is activated, all exposures are kept.
+                # To avoid coadding exposures, we store the forest with
+                # targetid_i (i = exposure index) instead of targetid
+                # The variable forests_by_targetid is modified by update_forest_dictionary.
+                if self.use_non_coadded_spectra & self.keep_single_exposures:
+                    # One exposure case
+                    if len(flux.shape) != 2:
+                        flux , ivar = np.array([flux]), np.array([ivar])
+                    for i, flux_i in enumerate(flux):
+                        forests_by_targetid, num_data = self.update_forest_dictionary(
+                                forests_by_targetid,
+                                f"{targetid}_{i}",
```

Comment:
Yes I totally agree with you. I started cross-exposure with no big hope. Now that we know that it work pretty well and that it can be used for other studies, I want to remove this hacky stuff.
But, it requires a lot of work indeed